### PR TITLE
Implement WalletConnect feature

### PR DESCRIPTION
**Pull request by skimaharvey**
Implemented WalletConnect version 1, as version 2 does not seem to work with our UP app.

### CODE REVIEW

File: components/UserInfos/UserInfos.tsx [hunk 1]
```diff
@@ -59,9 +60,16 @@ const UserInfos: React.FC<UserInfosProps> = ({ lsp3JSON }) => {
     setFirstLink(undefined);
   }, [lsp3JSON, address]);
 
+  const firstFourOfAddress = () => {
+    if (address) {
+      //firs two bytes of address
+      return address.slice(2, 6);
+    }
+  };
```
Comment (Hugoo) [inline components/UserInfos/UserInfos.tsx:68]: you can put this in utils file or smtgh other than this component as you may need to use this function elsewhere

File: components/UserInfos/UserInfos.tsx [hunk 2]
```diff
@@ -74,9 +82,18 @@ const UserInfos: React.FC<UserInfosProps> = ({ lsp3JSON }) => {
           alt="profileImage"
         />
         <div className="ml-4">
-          <div className="text-xs">{addressFormatter(address)}</div>
+          {/* <div className="text-xs">{addressFormatter(address)}</div> */}
           <div className="text-2xl font-bold my-2">
-            {profile?.LSP3Profile.name}
+            {profile?.LSP3Profile.name ? (
+              <>
```
Comment (Hugoo) [inline components/UserInfos/UserInfos.tsx:88]: do you really need the fragment here?

File: components/WalletConnector/WalletConnector.tsx [hunk 3]
```diff
@@ -0,0 +1,77 @@
+import QRCodeModal from '@walletconnect/qrcode-modal';
+import WalletConnect from '@walletconnect/client';
+import { useContext } from 'react';
+import { WalletAddressContext } from '../../contexts/WalletAddressContext';
+import { useRouter } from 'next/router';
+
+const WalletConnector = () => {
```
Comment (Hugoo) [inline components/WalletConnector/WalletConnector.tsx:7]:
```suggestion
const WalletConnector: React.FC = () => {
```

File: components/WalletConnector/WalletConnector.tsx [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+    // Subscribe to connection events
+    connector.on('connect', (error, payload) => {
+      if (error) {
+        throw error;
+      }
+
+      // Get provided accounts and chainId
+      const { accounts, chainId } = payload.params[0];
+      if (chainId != 2828) {
```
Comment (Hugoo) [inline components/WalletConnector/WalletConnector.tsx:30]: don't use 2828 here. It is not just a ramdom number, it is LUKSO L16 chain ID.

1) Put this number under a variable so we can **give a name** to this number.

chainId == LUKSO_L16_CHAIN_ID

is easier to understand than

chaindId == 2828

2) Put this in global conf file. You may need this number / variable in other parts of the code

File: components/WalletConnector/WalletConnector.tsx [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    });
+
+    connector.on('session_update', (error, payload) => {
+      if (error) {
+        throw error;
+      }
+
+      // Get updated accounts and chainId
+      const { accounts, chainId } = payload.params[0];
+      if (chainId != 2828) {
```
Comment (Hugoo) [inline components/WalletConnector/WalletConnector.tsx:45]: see previous comment. You reuse that same number as in line 30.

If tomorrow it changes to 56, you have to update it everywhere in your code. Use a variable.

File: package.json [hunk 6]
```diff
@@ -13,6 +13,9 @@
     "@erc725/erc725.js": "^0.14.4",
     "@lukso/lsp-smart-contracts": "^0.6.2",
     "@types/web3": "^1.2.2",
+    "@walletconnect/client": "^1.8.0",
+    "@walletconnect/qrcode-modal": "^1.8.0",
+    "@walletconnect/types": "^2.0.0-rc.2",
```
Comment (Hugoo) [inline package.json:18]: 2.0.0 or 1.8.0 ?

File: components/WalletConnector/WalletConnector.tsx (hunk 4 above)
Comment (Hugoo) [reply components/WalletConnector/WalletConnector.tsx:30]: It looks like you already have the variable

https://github.com/lukso-network/wallet.universalprofile.cloud/pull/12/files#diff-e5241839cc728bcc6af8c400f43bfa3f5da78ceceb24bfeff61cbab8e49e78ceR7

File: pages/connect-wallets.tsx [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+import { useRouter } from 'next/router';
+import { WalletAddressContext } from '../contexts/WalletAddressContext';
+import { toast } from 'react-toastify';
+import { L16_CHAIN_ID } from '../constants';
+
+const ConnectWallets: NextPage = () => {
+  const router = useRouter();
+  const { setWalletAddress } = useContext(WalletAddressContext);
+
+  const verifyChainId = () => {
```
Comment (Hugoo) [inline pages/connect-wallets.tsx:13]: maybe `isValidChainId` is a better name?

File: pages/connect-wallets.tsx [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    const chainId = parseInt(window.ethereum.networkVersion);
+    if (chainId && window.ethereum.networkVersion != L16_CHAIN_ID) {
+      toast('Please switch to L16 network');
+      return false;
+    } else {
+      return true;
+    }
```
Comment (Hugoo) [inline pages/connect-wallets.tsx:20]:
```suggestion
    if (chainId && window.ethereum.networkVersion != L16_CHAIN_ID) {
      toast('Please switch to L16 network');
      return false;
    }
    
    return true;
```

you return in the if so no need to add a else. The linter / or TS should tell you

File: pages/connect-wallets.tsx [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+    try {
+      const { ethereum } = window;
+      if (!verifyChainId()) return;
+      await ethereum
+        .request({
+          method: 'eth_requestAccounts',
+        })
+
+        .then(function (accounts: string[]) {
```
Comment (Hugoo) [inline pages/connect-wallets.tsx:37]: you await AND then ?

Why both?

Either do it in await or do it in .then() no?

**Review by Hugoo — CHANGES_REQUESTED** [review]

File: components/UserInfos/UserInfos.tsx (hunk 2 above)
Comment (skimaharvey) [reply components/UserInfos/UserInfos.tsx]: I do need to check this condition but I created a function to do it above

File: components/SideBar/SideBar.tsx [hunk 10]
```diff
@@ -30,52 +33,37 @@ const SideBar: React.FC = () => {
...
-      } else {
-        setWalletAddress('');
+        return;
       }
+      setWalletAddress('');
     });
 
     ethereum.on('chainChanged', function (networkId: number) {
-      if (networkId != 2828) {
+      if (networkId != L16_CHAIN_ID) {
```
Comment (Hugoo) [inline components/SideBar/SideBar.tsx:66]: !==

File: components/SideBar/SideBar.tsx [hunk 11]
```diff
@@ -93,7 +81,7 @@ const SideBar: React.FC = () => {
         Connect to the UP extension to manage your portfolio
       </div>
       <button
-        onClick={loginExtension}
+        onClick={() => router.push('/connect-wallets')}
         className="flex items-center rounded-lg bg-deepPink py-1 px-4"
```
Comment (Hugoo) [inline components/SideBar/SideBar.tsx:85]: I suggest to use Next Link instead: https://nextjs.org/docs/api-reference/next/link

File: components/UserInfos/UserInfos.tsx [hunk 12]
```diff
@@ -46,8 +48,22 @@ const UserInfos: React.FC<UserInfosProps> = ({ lsp3JSON }) => {
...
+      return (
+        <>
+          <span>{profile?.LSP3Profile.name}</span>
+          <span className="text-sm ml-2 text-gray-400">
+            #{firstTwoBytesOfAddress(address)}
+          </span>
+        </>
+      );
+    }
+    return <>Anonymous</>;
```
Comment (Hugoo) [inline components/UserInfos/UserInfos.tsx:62]: Try to use less Fragments

File: components/WalletConnector/WalletConnector.tsx [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+    });
+  };
+
+  return (
+    <div
+      className="flex flex-col items-center py-4 cursor-pointer"
+      onClick={() => showModal()}
+    >
+      <img
+        src="./WalletConnect-icon.png"
```
Comment (Hugoo) [inline components/WalletConnector/WalletConnector.tsx:67]: Please organise better the public folder (you can put images in a img/ folder for instance). When we will add more things we will be happy to have them organised. - You can also remove the vercel.svg image

File: components/WalletConnector/WalletConnector.tsx [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+        alt="walletConnect image"
```
Comment (Hugoo) [inline components/WalletConnector/WalletConnector.tsx:68]: WalletConnect logo

File: hooks/useWalletConnect.ts [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+import { L16_CHAIN_ID } from '../constants';
+import { WalletAddressContext } from '../contexts/WalletAddressContext';
+import walletConnectConnector from '../utils/walletConnectConnector';
+import { useRouter } from 'next/router';
+
+const useWalletConnect = () => {
+  const { setWalletAddress } = useContext(WalletAddressContext);
+
+  const router = useRouter();
+  let account;
```
Comment (Hugoo) [inline hooks/useWalletConnect.ts:11]: You can probably type this as a string.

File: hooks/useWalletConnect.ts [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+
+    // Subscribe to connection events
+    connector.on('connect', (error, payload) => {
+      if (error) {
+        throw error;
+      }
+
+      // Get provided accounts and chainId
+      const { accounts, chainId } = payload.params[0];
+      if (chainId != L16_CHAIN_ID) {
```
Comment (Hugoo) [inline hooks/useWalletConnect.ts:29]: !==

File: hooks/useWalletConnect.ts [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+    connector.on('disconnect', (error) => {
+      if (error) {
+        throw error;
+      }
+
+      // Delete connector
+      setWalletAddress('');
+    });
+  }, []);
+  return account;
```
Comment (Hugoo) [inline hooks/useWalletConnect.ts:61]: TS says it is undefined

File: utils/firstTwoBytesOfAddress.ts [hunk 18]
```diff
@@ -0,0 +1,7 @@
+const firstTwoBytesOfAddress = (address: string) => {
```
Comment (Hugoo) [inline utils/firstTwoBytesOfAddress.ts:1]: maybe not worth using a single file for this very little function.
Maybe just create an utils.ts file so you can put a few more functions?

File: utils/walletConnectConnector.ts [hunk 19]
```diff
@@ -0,0 +1,12 @@
+import WalletConnect from '@walletconnect/client';
+import QRCodeModal from '@walletconnect/qrcode-modal';
+
+const walletConnectConnector = () => {
+  const connector = new WalletConnect({
+    bridge: 'https://bridge.walletconnect.org', // Required
+    qrcodeModal: QRCodeModal,
+  });
+  return connector;
+};
```
Comment (Hugoo) [inline utils/walletConnectConnector.ts:10]:
```suggestion

const WALLET_CONNECT_BRIDGE_URL = 'https://bridge.walletconnect.org';

const walletConnectConnector = () => {
  const connector = new WalletConnect({
    bridge: WALLET_CONNECT_BRIDGE_URL, // Required
    qrcodeModal: QRCodeModal,
  });
  return connector;
};
```

**Review by Hugoo — APPROVED** [review]

**Pull request merged**